[TF:TRT] refactor einsum converter

christopherbate opened the pull request:
This change moves the conversion function for the "Einsum" operation and all associated helper functions to a separate translation unit `tf2tensorrt/convert/ops/einsum.cc`. A small amount of refactoring is applied to eliminate functions which duplicate the functionality found in `TRTNetworkBuilder`. Functions which exceed a certain "cognitive complexity" calculated by clang are refactored to be below the threshold. Tests are updated to reflect the fact that certain error messages changed when updating the converter to use the `OpConverterBase` class.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:
```diff
@@ -0,0 +1,721 @@
+}
+
+// Layout of the einsum dimensions: Batch, Free and Contraction indices.
+// Example: abcd,adef -> abde. The first tensor has layout BFC, the second BCF.
```
I understand this is from the old code, but when I read this again, I don't feel it is correct. Can you double check?

here batch: ad, free:be, contract:cf. right?
may we need to change the expression to adbc,adfe -> abde
to make the remaining part of the sentence correct?

Can you explain if I am not right here?

christopherbate replied:
Right I think the comment is not correct, will fix

christopherbate replied:
applied

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:
```diff
@@ -0,0 +1,721 @@
+  TRT_ENSURE(operand.is_weights());
+  if (!need_transpose) {
+    // If we don't need to transpose, then the operand remains as a weights
+    // constant. In this case we also don't need a reshape?
```
s/?/./

tfeher reviewed (commented):
Thanks @christopherbate for this PR! It looks good to me, just one suggestion: enable the last unit test.

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:
```diff
@@ -0,0 +1,721 @@
+template <typename Iterator1, typename Iterator2>
+Status FindIndices(Iterator1 begin, Iterator1 end, Iterator2 array_begin,
+                   Iterator2 array_end, std::vector<int>* permute) {
+  const int n = array_end - array_begin;
+  if (n < end - begin) {
+    return errors::Internal("Incorrect array size");
+  }
```
Reworked to make name/params explicit and accept Span, which is slightly cleaner:

```
template <typename T>
Status FindIndicesoOfAllValuesInSrc(absl::Span<const T> values,
                                    absl::Span<const T> src,
                                    std::vector<int>* indices) {
  const int n = values.size();
  if (src.size() < values.size()) {
    return errors::Internal(
        "Span 'src' cannot contain all elements of 'values'");
  }
  for (auto i = 0; i < values.size(); i++) {
    auto iter = absl::c_find(src, values[i]);
    if (iter == src.end()) {
      return errors::Internal("Label ", values[i], " not found");
    }
    int idx = std::distance(src.begin(), iter);
    indices->push_back(idx);
  }
  return Status::OK();
}
```

christopherbate commented on tensorflow/compiler/tf2tensorrt/convert/utils.h:
```diff
@@ -190,6 +191,27 @@ Status TrtDimsToTensorShape(const nvinfer1::Dims trt_dims,
+}
+
+template <typename T>
+T TrtDimsVolumeUnsafe(const nvinfer1::Dims& dims, int32_t offset = 0,
```
I think it was more because it couldn't handle some "scalar shape" edge case. I'm going to rebase this on the PR for adding `DimsAdapter` since that one will merge before this and these functions go away.

christopherbate committed:
[TF:TRT] Refactor einsum converter

[TF:TRT] Make Einsum converter use DimsAdapter

[TF:TRT] Reactivate einsum test case for TRT >= 8.0

christopherbate commented:
Rebased on #53602, which should be merged shortly. Added 1 commit to use `DimsAdapter` to address comment.

bixia1 commented:
Can you squash the commits into one? thanks!

christopherbate commented:
squashed

bixia1 commented:
Would you please fix these errors:
[third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:15](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc?l=15&ws=tap-prod-presubmit/170220345&snapshot=2):19: error: extra tokens at end of #ifdef directive [-Werror,-Wextra-tokens]
#ifdef GOOGLE_CUDA&& GOOGLE_TENSORRT
                  ^
                  //
[third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:46](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc?l=46&ws=tap-prod-presubmit/170220345&snapshot=2):13: error: unused variable 'n' [-Werror,-Wunused-variable]
  const int n = values.size();
            ^
[third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:681](https://cs.corp.google.com/piper///depot/google3/third_party/tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc?l=681&ws=tap-prod-presubmit/170220345&snapshot=2):17: error: unused variable 'node_def' [-Werror,-Wunused-variable]

christopherbate commented:
Rebased and squashed the fixes for clang warnings.

christopherbate commented:
I guess I probably didn't fix the clang warning regarding the preprocessor. I didn't realize the issue is probably because of writing `#ifdef` instead of `#if` until you commented on my other PR. Let me know if you changed this on your end or if I need to push the change

bixia1 commented:
> I guess I probably didn't fix the clang warning regarding the preprocessor. I didn't realize the issue is probably because of writing `#ifdef` instead of `#if` until you commented on my other PR. Let me know if you changed this on your end or if I need to push the change

Would you please fix that and squash?

christopherbate commented:
done
